scripts/error-docgen.mjs

```diff
@@ -0,0 +1,256 @@
+			'\n',
+		]
+			.filter((l) => l !== undefined)
+			.join('\n');
+	}
+
+	let astroResult = '';
+	for (const comment of astroErrorData.jsdoc) {
+		if (comment.kind === 'heading') {
+			astroResult += `## ${comment.name}\n\n`;
+			if (comment.description) {
+				astroResult += comment.description.trim() + '\n\n';
+			}
+			continue;
+		}
+
+		if (comment.see && comment.see.length === 1) {
+			comment.see = comment.see[0].split('\n');
+		}
+
+		const cleanMessage = comment.tags.find((tag) => tag.title === 'message')?.value;
+		astroResult += [
+			`### ${sanitizeString(
+				astroErrorData.errors[comment.meta.code.name].title ?? comment.longname
+			)}`,
+			comment.deprecated
+				? [
+						``,
+						':::caution[Deprecated]',
+						typeof comment.deprecated === 'string'
+							? comment.deprecated
+							: 'This error cannot be emitted by Astro anymore',
+						':::',
+				  ]
+						.filter((l) => l !== undefined)
+						.join('\n')
+				: undefined,
+			``,
+			`> ${getMessage(
+				comment.longname,
+				astroErrorData.errors[comment.longname].message,
+				cleanMessage
+			)} (E${padCode(astroErrorData.errors[comment.longname].code)})\n`,
+			comment.description && `#### What went wrong?`,
+			comment.description && comment.description.trim(),
+			``,
+			comment.see
+				? `**See Also:**\n${comment.see.map((s) => `- ${s.replace('-', '')}`.trim()).join('\n')}`
+				: undefined,
+			`\n\n`,
+		]
+			.filter((l) => l !== undefined)
+			.join('\n');
+	}
+
+	compilerResult = compilerResult.replace(/https\\?:\/\/docs\.astro\.build\//g, '/');
+	astroResult = astroResult.replace(/https\\?:\/\/docs\.astro\.build\//g, '/');
+
+	fs.writeFileSync(
+		'src/pages/en/reference/error-reference.md',
+		HEADER + compilerResult + astroResult + FOOTER,
+		'utf8'
+	);
+
+	function getMessage(errorName, message, cleanMessage) {
+		if (cleanMessage) {
+			return cleanMessage;
+		}
+
+		if (message) {
+			if (typeof message === 'string') {
+				return `**${errorName}**: ${message}`;
+			} else {
+				return `**${errorName}**: ${String.raw({
+					raw: extractStringFromFunction(message.toString()),
+				})}`;
+			}
+		}
+
+		return undefined;
+	}
+}
+
+async function getAstroErrorsData() {
+	const inputBuffer = STUB || (await fetch(errorURL).then((r) => r.text()));
+
+	const compiledResult = ts.transpileModule(inputBuffer, {
+		compilerOptions: { module: 'esnext', target: 'esnext', removeComments: false },
+	}).outputText;
+
+	const encodedJs = encodeURIComponent(compiledResult);
+	const dataUri = 'data:text/javascript;charset=utf-8,' + encodedJs;
+
+	/**
+	 * @type {{AstroErrorData: Object.<string, {code: number, message: string, hint: string}>}
+	 */
+	const data = await import(dataUri);
+
+	const jsDocComments = jsdoc
+		.explainSync({ source: compiledResult })
+		.filter(
+			(data) =>
+				data.kind !== 'package' &&
+				!data.undocumented &&
+				data.tags &&
+				data.tags.some((tag) => tag.title === 'docs')
+		);
```

Comment:
This logic is used here and in L191 and I’m not really able to follow what’s going on just by reading it. Might be good to either extract to a reused function with a descriptive name or just add a comment to each of them that explains exactly what’s being filtered for and why.